web/src/App.tsx

```diff
@@ -90,161 +85,156 @@ export const theme = createMuiTheme({
   },
 });
 
-class App extends React.Component<{}, State> {
-  constructor(props: {}) {
-    super(props);
-    this.state = { songs: [], showAbout: false };
-  }
+const App = () => {
+  const [songs, setSongs] = useState<Song[]>([]);
+  const [showingAbout, showAbout] = useState(false);
 
-  componentDidMount() {
+  useEffect(() => {
     fetch(songUrl)
       .then((res) => {
         return res.text();
       }).then((body) => {
         const lines = body.split("\n");
-        this.setState({
-          songs: lines.slice(2, lines.length).map(line => {
-            const elements = line.split("\t");
-            return {
-              url: elements[0],
-              title: elements[1],
-              publishedAt: new Date(elements[2]),
-              thumbnailUrl: elements[3],
-              songType: elements[5],
-              originalSongTitle: elements[6],
-              originalArtist: elements[7],
-              originalCreator: elements[8],
-              tag: elements[9],
-            };
-          }).sort((a, b) => {
-            // const songType = (songMap.get(a.songType) || 9) - (songMap.get(b.songType) || 9);
-            // if (songType !== 0) return songType;
+        const songs = lines.slice(2, lines.length).map(line => {
+          const elements = line.split("\t");
+          return {
+            url: elements[0],
+            title: elements[1],
+            publishedAt: new Date(elements[2]),
+            thumbnailUrl: elements[3],
+            songType: elements[5],
+            originalSongTitle: elements[6],
+            originalArtist: elements[7],
+            originalCreator: elements[8],
+            tag: elements[9],
+          };
+        }).sort((a, b) => {
+          // const songType = (songMap.get(a.songType) || 9) - (songMap.get(b.songType) || 9);
+          // if (songType !== 0) return songType;
 
-            const publishedAt = b.publishedAt.getTime() - a.publishedAt.getTime();
-            return publishedAt;
-          }),
+          const publishedAt = b.publishedAt.getTime() - a.publishedAt.getTime();
+          return publishedAt;
         });
+
+        setSongs(songs);
       });
-  }
+  });
```

Comment:
Thank you, great pull request.

However, it seems that useEffect is called repeatedly. Passing an empty array as the second argument seems to solve the problem.

- [Tip: Optimizing Performance by Skipping Effects | Using the Effect Hook – React](https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects)

```suggestion
  }, []);
```

Reply:
Oh thx!